Readme.md
ical2org.py
===========

This script converts an ical calendar (for instance, as exported from google
calendar) into an org-mode document. It is conceived as a replacement of the
awk script located here:

http://orgmode.org/worg/org-tutorials/org-google-sync.html

The main difference is that ical2org.py correctly manages recurring events
of "yearly", "daily" and "weekly" types. ical2org.py duplicates all
recurring events falling into a specified time-frame into the exported
org-document.

Installation
===========

ical2org.py is a python script and it needs the following libraries:

- icalendar
- pytz
- tzlocal (optional)

You can install those packages as follows (thanks to Guido Van Hoecke):

Get iCalendar from https://pypi.python.org/pypi/icalendar, untar it and run
(you probably will also need the "python-setuptools" package)

````shell
sudo python setup.py install
````

Get pytz from https://pypi.python.org/pypi/pytz/. It comes as an egg
file so you just run 

````shell
sudo easy_install pytz-2013b-py2.7.egg
````

Customize the script
====================

If you have `tzlocal` installed, the script will use the system
timezone. Otherwise, You should edit the script and modify the lines at the
beginning in order to specify your local timezone (line 9) and the window of
the time-frame, in days (the time-frame is relative to the current time). You
can also modify the org TAG used for specifying that an event is recurring.


CHANGELOG.rst
CHANGELOG
=========

0.5
---
* Fixed bug with whole day events with times at midnight

0.4
---
* Rewrote a lot of functionality using recurring-ical-events package
* All day events are in UTC. Fixes #42
* manage events with single attendees (closes #37)
* Manage declined events (closes #36), using ``-e``
* Added ``--continue-on-error`` flag
* Many other bug fixes

0.3
---

* Support excluding dates EXDATE (#32)
* Remove misleading info of package not being in PyPI (#31)
* Include location data, if present. Closes #30
* catch errors when end datetime of event is missing
* fix #28 by also checking datetimes for missing timezone info

0.2
---

* Add license and small fixes (mailmap)

0.1
---

* Slightly modify README to include installation from git source
* fix the "TypeError: 'range' object does not support item deletion"
* Removed import of unicode\_literals as it is discouraged (by click)
* Made py2 and py3 compatible (tested 2.7, 3.5, 3.6)
* Removed extraneous prints from testing
* Added test suite (pytest based)
* CLI by click: added --timezone, --days and --print-timezones
* Readme.md updated and renamed to README.rst. tox.ini updated
* Refactored to Convertor class
* Code reformatted, fixed some exception handling
* Print warning, if cannot import tzlocal
* Script refactored to main and convert functions
* Rewrite from script to python package
* Add GPL v3 license
* Use system timezone via  tzlocal
* Handle zero length events
* \* Proper printout of all day events
* Die gracefully if ical file is corrupt
* README: Little change in installation instructions
* ! indentations
* ! remove unnecessary escapes from SUMMARY
* Update Readme.md
* Update ical2org.py
* ! take care of unnecessary escapes and newline for DESCRIPTION
* write to the output file when provided
* Do not use EventRecurDailyIter/EventRecurWeeklyIter. Use EventRecurDaysIter instead
* fix bug
* fix little bug
* Complete rewriting of ical2org using iterators
* last day of daily repeat was being omitted
* Fix end time of recurring events
* include descriptions in the items
* Add instructions for installing dependencies
* get input from STDIN if no filename given
* Deal with yearly recurring events
* add .gitignore
* Add Readme
* put main loop under a try block
* do not crash on monthly or yearly recurring events
* Store interval time following UTC
* Fix bug. Add date normalizations where needed
* Add user-defined parameters for window and tag
* Properly deal with dates in other timezones
* do not use utc for canonical date
* Take timezones into account
* Add recurring tag to recurring events
* first version, seems to work
* init commit


tests/test_datetime.py
"""Test functions related to simple time conversions.
"""
from datetime import datetime

import pytest
from pytz import timezone

from ical2orgpy import org_datetime, org_date

# Timezone in Prague
PRAGUE = timezone("Europe/Prague")
# UTC
UTC = timezone("UTC")


@pytest.mark.parametrize(
    "dt, tz, expected", [
        (datetime(2017, 12, 15, 17, 35, 0, 0, UTC), PRAGUE,
         "<2017-12-15 Fri 18:35>"),
        (datetime(2017, 12, 15, 18, 35, 0, 0, UTC), PRAGUE,
         "<2017-12-15 Fri 19:35>"),
    ],
    ids=lambda itm: str(itm))
def test_org_datetime(dt, tz, expected):
    res = org_datetime(dt, tz)
    assert res == expected


@pytest.mark.parametrize(
    "dt, tz, expected", [
        (datetime(2017, 12, 15, 17, 35, 0, 0, UTC), PRAGUE,
         "<2017-12-15 Fri>"),
        (datetime(2017, 12, 15, 18, 35, 0, 0, UTC), PRAGUE,
         "<2017-12-15 Fri>"),
    ],
    ids=lambda itm: str(itm))
def test_org_date(dt, tz, expected):
    res = org_date(dt, tz)
    assert res == expected


tests/test_arithmetics.py
from datetime import datetime, timedelta

import pytest

from pytz import timezone, utc

from ical2orgpy import add_delta_dst, advance_just_before

# Timezone in Prague
PRAGUE = timezone("Europe/Prague")
# UTC
UTC = timezone("UTC")


# TODO: Add some cases spanning DST borders
@pytest.mark.parametrize(
    "dt, delta, expected",
    [(datetime(2017, 12, 15, 12, 0, 0, 0, PRAGUE), timedelta(hours=1),
      datetime(2017, 12, 15, 13, 0, 0, 0, PRAGUE)),
     (datetime(2017, 12, 15, 12, 0, 0, 0, PRAGUE), timedelta(days=1),
      datetime(2017, 12, 16, 12, 0, 0, 0, PRAGUE))],
    ids=lambda itm: str(itm))
def test_add_delta_dst(dt, delta, expected):
    res = add_delta_dst(dt, delta)
    assert isinstance(res, datetime)
    assert hasattr(res, "tzinfo")
    delta = (res - expected).total_seconds()
    # tolerate 5 minutes difference (yes, it happens)
    assert abs(delta) < 5 * 60


@pytest.mark.parametrize("start_dt, timeframe_start, delta_days, expected",
                         [(datetime(2003, 2, 1, 1, 0, 0, 0, PRAGUE),
                           datetime(2017, 9, 16, 18, 57, 53, 0, PRAGUE), 7,
                           (datetime(2017, 9, 9, 1, 0, 0, 0, PRAGUE), 762))],
                         ids=lambda itm: str(itm))
def test_advance_just_before(start_dt, timeframe_start, delta_days, expected):
    res = advance_just_before(start_dt, timeframe_start, delta_days)
    res_dt, res_int = res
    exp_dt, exp_int = expected
    assert res_int == exp_int
    assert res_dt.year == exp_dt.year
    assert res_dt.month == exp_dt.month
    assert res_dt.day == exp_dt.day
    assert res_dt.hour == exp_dt.hour
    assert res_dt.minute == exp_dt.minute


tests/filters/test_filters.py
from datetime import datetime
import pytest
from py.path import local
import yaml
from freezegun import freeze_time
from pytz import timezone

from ical2orgpy import Convertor

FILEROOT = local("tests/filters/scenarios")

assert FILEROOT.exists()


class Scenario(object):
    @classmethod
    def discover(cls, fileroot):
        return [Scenario(cfgfile) for cfgfile in fileroot.listdir("*.yaml")]

    def __init__(self, cfg_file):
        self.cfg_file = cfg_file
        self.fileroot = cfg_file.dirpath()
        """directory where are all the scenarios"""

        self.name = cfg_file.purebasename
        ics_purebasename = self.name.rsplit(".", 1)[0]
        self.ics_file = cfg_file.new(purebasename=ics_purebasename, ext=".ics")
        """ICS file serving as conversion input"""

        assert self.ics_file, "ICS file exists"

        self.expected_file = cfg_file.new(ext=".expected.org")
        """Expected result of conversion to org format"""
        assert self.expected_file.exists()

        with cfg_file.open("r", encoding="utf-8") as f:
            self.cfg = yaml.load(f, Loader=yaml.FullLoader)

        self.desciption = self.cfg["description"]
        """Test case description"""

        # YAML converts ctime to datetime
        self.context = self.cfg["context"]

        # no conversions here
        self.parameters = self.cfg["parameters"]

    def freeze_time(self):
        """Provide context manager mocking context time from scenario.

        Using freezegun utility, it temporarily provide different
        "now" time for datetime based functions.
        """
        ctime = self.context["ctime"]
        tz = timezone(self.context["tzone"])
        is_dst = self.context.get("is_dst", False)
        offset = tz.utcoffset(ctime, is_dst=is_dst).total_seconds()
        return freeze_time(ctime, tz_offset=offset / 3600)


SCENARIOS = Scenario.discover(FILEROOT)


@pytest.mark.parametrize("scenario", SCENARIOS, ids=lambda itm: itm.name)
def test_conversion(scenario, tmpdir):
    cfg_file = tmpdir / "cfg_file.yaml"
    ics_file = tmpdir / "input.ics"
    expected_file = tmpdir / "expected.org"

    # copy files from scenario into tmpdir
    # it shall make researching problems simpler
    scenario.cfg_file.copy(cfg_file)
    scenario.ics_file.copy(ics_file)
    scenario.expected_file.copy(expected_file)

    assert cfg_file.exists()
    assert ics_file.exists()
    assert expected_file.exists()

    convertor = Convertor(scenario.parameters["days"],
                          scenario.parameters["tzone"],
                          scenario.parameters["email"])

    result_file = tmpdir / "output.org"
    with scenario.freeze_time():
        with ics_file.open("r", encoding="utf-8") as ics_f:
            with result_file.open("w", encoding="utf-8") as result_f:
                convertor(ics_f, result_f)
    assert result_file.exists()
    res_txt = result_file.read_text(encoding="utf-8")
    exp_txt = expected_file.read_text(encoding="utf-8")
    assert res_txt == exp_txt


ical2org.py
#!/usr/bin/python

import sys
from math import floor
from datetime import date, datetime, timedelta, tzinfo
from icalendar import Calendar
from pytz import timezone, utc

try:
    from tzlocal import get_localzone
    LOCAL_TZ = get_localzone()
except:
    # Change here your local timezone
    LOCAL_TZ = timezone("Europe/Paris")

# Window length in days (left & right from current time). Has to be positive.
WINDOW = 90
# leave empty if you don't want to attach any tag to recurring events
RECUR_TAG = ":RECURRING:"

# Do not change anything below

def orgDatetime(dt):
    '''Given a datetime in his own timezone, return YYYY-MM-DD DayofWeek HH:MM in local timezone'''
    return dt.astimezone(LOCAL_TZ).strftime("<%Y-%m-%d %a %H:%M>")

def orgDate(dt):
    '''Given a date in his own timezone, return YYYY-MM-DD DayofWeek in local timezone'''
    return dt.astimezone(LOCAL_TZ).strftime("<%Y-%m-%d %a>")

def get_datetime(dt):
    '''Given a datetime, return it. If argument is date, convert it to a local datetime'''
    if isinstance(dt, datetime):
        return dt
    else:
        # d is date. Being a naive date, let's suppose it is in local
        # timezone.  Unfortunately using the tzinfo argument of the standard
        # datetime constructors ''does not work'' with pytz for many
        # timezones, so create first a utc datetime, and convert to local timezone
        aux_dt = datetime(year = dt.year, month = dt.month, day = dt.day, tzinfo = utc)
        return aux_dt.astimezone(LOCAL_TZ)

def add_delta_dst(dt, delta):
    '''Add a timedelta to a datetime, adjusting DST when appropriate'''
    # convert datetime to naive, add delta and convert again to his own timezone
    naive_dt = dt.replace(tzinfo = None)
    return dt.tzinfo.localize(naive_dt + delta)

def advance_just_before(start_dt, timeframe_start, delta_days):
    '''Advance an start_dt datetime to the first date just before
    timeframe_start. Use delta_days for advancing the event. Precond:
    start_dt < timeframe_start'''
    delta = timedelta(days = delta_days)
    delta_ord = floor( (timeframe_start.toordinal() - start_dt.toordinal() - 1) / delta_days )
    return (add_delta_dst(start_dt, timedelta(days = delta_days * int(delta_ord))), int(delta_ord))


def generate_event_iterator(comp, timeframe_start, timeframe_end):
    ''' Given an VEVENT object return an iterator with the proper delta (days, weeks, etc)'''
    # Note: timeframe_start and timeframe_end are in UTC
    if comp.name != 'VEVENT': return []
    if 'RRULE' in comp:
        return {
            'WEEKLY' : EventRecurDaysIter(7, comp, timeframe_start, timeframe_end),
            'DAILY' : EventRecurDaysIter(1, comp, timeframe_start, timeframe_end),
            'MONTHLY' : [],
            'YEARLY' : EventRecurYearlyIter(comp, timeframe_start, timeframe_end)
            }[ comp['RRULE']['FREQ'][0] ]
    else:
        return EventSingleIter(comp, timeframe_start, timeframe_end)

class EventSingleIter:
    '''Iterator for non-recurring single events.'''
    def __init__(self, comp, timeframe_start, timeframe_end):
        self.ev_start = get_datetime(comp['DTSTART'].dt)

        # Events with the same begin/end time same do not include
        # "DTEND".
        if "DTEND" not in comp:
            self.ev_end = self.ev_start
        else:
            self.ev_end = get_datetime(comp['DTEND'].dt)

        self.duration = self.ev_end - self.ev_start
        self.result = ()
        if (self.ev_start < timeframe_end and self.ev_end > timeframe_start):
            self.result = ( self.ev_start, self.ev_end, 0)

    def __iter__(self):
        return self

    # Iterate just once
    def __next__(self):
        if self.result:
            aux = self.result
            self.result = ()
        else:
            raise StopIteration
        return aux

class EventRecurDaysIter:
    '''Iterator for daily-based recurring events (daily, weekly).'''
    def __init__(self, days, comp, timeframe_start, timeframe_end):
        self.ev_start = get_datetime(comp['DTSTART'].dt)
        self.ev_end = get_datetime(comp['DTEND'].dt)
        self.duration = self.ev_end - self.ev_start
        self.is_count = False
        if 'COUNT' in comp['RRULE']:
            self.is_count = True
            self.count = comp['RRULE']['COUNT'][0]
        delta_days = days
        if 'INTERVAL' in comp['RRULE']:
            delta_days *= comp['RRULE']['INTERVAL'][0]
        self.delta = timedelta(delta_days)
        if 'UNTIL' in comp['RRULE']:
            if self.is_count:
                raise "UNTIL and COUNT MUST NOT occur in the same 'recur'"
            self.until_utc = get_datetime(comp['RRULE']['UNTIL'][0]).astimezone(utc)
        else :
            self.until_utc = timeframe_end
        if self.until_utc < timeframe_start:
            self.current = self.until_utc + self.delta # Default value for no iteration
            return
        self.until_utc = min(self.until_utc, timeframe_end)
        if self.ev_start < timeframe_start:
            # advance to timeframe start
            (self.current, counts) = advance_just_before(self.ev_start, timeframe_start, delta_days)
            if self.is_count:
                self.count -= counts
                if self.count < 1: return
            while self.current < timeframe_start:
                self.current = add_delta_dst(self.current, self.delta)
        else:
            self.current = self.ev_start

    def __iter__(self):
        return self

    def next_until(self):
        if self.current > self.until_utc:
            raise StopIteration
        event_aux = self.current
        self.current = add_delta_dst(self.current, self.delta)
        return (event_aux, event_aux.tzinfo.normalize(event_aux + self.duration), 1)

    def next_count(self):
        if self.count < 1:
            raise StopIteration
        self.count -= 1
        event_aux = self.current
        self.current = add_delta_dst(self.current, self.delta)
        return (event_aux, event_aux.tzinfo.normalize(event_aux + self.duration), 1)

    def next(self):
        if self.is_count: return self.next_count()
        return self.next_until()

class EventRecurMonthlyIter:
    pass

class EventRecurYearlyIter:
    def __init__(self, comp, timeframe_start, timeframe_end):
        self.ev_start = get_datetime(comp['DTSTART'].dt)
        self.ev_end = get_datetime(comp['DTEND'].dt)
        self.start = timeframe_start
        self.end = timeframe_end
        self.is_until = False
        if 'UNTIL' in comp['RRULE']:
            self.is_until = True
            self.end = min(self.end, get_datetime(comp['RRULE']['UNTIL'][0]).astimezone(utc))
        if self.end < self.start:
            # Default values for no iteration
            self.i = 0
            self.n = 0
            return
        if 'BYMONTH' in comp['RRULE']:
            self.bymonth = comp['RRULE']['BYMONTH'][0]
        else:
            self.bymonth = self.ev_start.month
        if 'BYMONTHDAY' in comp['RRULE']:
            self.bymonthday = comp['RRULE']['BYMONTHDAY'][0]
        else:
            self.bymonthday = self.ev_start.day
        self.duration = self.ev_end - self.ev_start
        self.years = range(self.start.year, self.end.year + 1)
        if 'COUNT' in comp['RRULE']:
            if self.is_until:
                raise "UNTIL and COUNT MUST NOT occur in the same 'recur'"
            self.years = range(self.ev_start.year, self.end.year + 1)
            del self.years[comp['RRULE']['COUNT'][0]:]
        self.i = 0
        self.n = len(self.years)

    def __iter__(self):
        return self

    def next(self):
        if self.i >= self.n: raise StopIteration
        event_aux = self.ev_start.replace(year = self.years[self.i])
        event_aux = event_aux.replace(month = self.bymonth)
        event_aux = event_aux.replace(day = self.bymonthday)
        self.i = self.i + 1;
        if event_aux > self.end: raise StopIteration
        if event_aux < self.start: return self.next()
        return (event_aux, event_aux.tzinfo.normalize(event_aux + self.duration), 1)

if len(sys.argv) < 2:
    fh = sys.stdin
else:
    fh = open(sys.argv[1], 'r', encoding='utf-8')

if len(sys.argv) > 2:
    fh_w = open(sys.argv[2], 'w', encoding='utf-8')
else:
    fh_w = sys.stdout

try:
    cal = Calendar.from_ical(fh.read())
except:
    print("ERROR parsing ical file", file=sys.stderr)
    exit(1)
    pass

now = datetime.now(utc)
start = now - timedelta( days = WINDOW)
end = now + timedelta( days = WINDOW)
for comp in cal.walk():
    try:
        event_iter = generate_event_iterator(comp, start, end)
        for comp_start, comp_end, rec_event in event_iter:
            SUMMARY = ""
            if "SUMMARY" in comp:
                SUMMARY = comp['SUMMARY'].to_ical().decode(encoding='utf-8')
                SUMMARY = SUMMARY.replace('\\,', ',')
            if not len(SUMMARY):
                SUMMARY = "(No title)"
            fh_w.write("* {}".format(SUMMARY))
            if rec_event and len(RECUR_TAG):
                fh_w.write(" {}\n".format(RECUR_TAG))
            fh_w.write("\n")
            if isinstance(comp["DTSTART"].dt, datetime):
                fh_w.write("  {}--{}\n".format(orgDatetime(comp_start), orgDatetime(comp_end)))
            else:  # all day event
                fh_w.write("  {}--{}\n".format(orgDate(comp_start), orgDate(comp_end - timedelta(days=1))))
            if 'DESCRIPTION' in comp:
                DESCRIPTION = '\n'.join(comp['DESCRIPTION'].to_ical().split('\\n'))
                DESCRIPTION = DESCRIPTION.replace('\\,', ',')
                fh_w.write("{}\n".format(DESCRIPTION))

            fh_w.write("\n")
    except:
        pass

exit();


ical2orgpy.py
from __future__ import print_function

import sys
import traceback
from datetime import date, datetime, timedelta, time

import click
import recurring_ical_events
from icalendar import Calendar
from pytz import all_timezones, timezone, utc
from tzlocal import get_localzone

MIDNIGHT = time(0, 0, 0)


def org_datetime(dt, tz):
    '''Timezone aware datetime to YYYY-MM-DD DayofWeek HH:MM str in localtime.
    '''
    return dt.astimezone(tz).strftime("<%Y-%m-%d %a %H:%M>")


def org_date(dt, tz):
    '''Timezone aware date to YYYY-MM-DD DayofWeek in localtime.
    '''
    if hasattr(dt, "astimezone"):
        dt = dt.astimezone(tz)
    return dt.strftime("<%Y-%m-%d %a>")


def event_is_declined(comp, emails):
    attendee_list = comp.get('ATTENDEE', None)
    if attendee_list:
        if not isinstance(attendee_list, list):
            attendee_list = [attendee_list]
        for att in attendee_list:
            if att.params.get('PARTSTAT', '') == 'DECLINED' and att.params.get('CN', '') in emails:
                return True
    return False


class IcalError(Exception):
    pass


class Convertor():
    RECUR_TAG = ":RECURRING:"

    # Do not change anything below

    def __init__(self, days=90, tz=None, emails=[], include_location=True, continue_on_error=False):
        """
        days: Window length in days (left & right from current time). Has
        to be positive.
        tz: timezone. If None, use local timezone.
        emails: list of user email addresses (to deal with declined events)
        """
        self.emails = set(emails)
        self.tz = timezone(tz) if tz else get_localzone()
        self.days = days
        self.include_location = include_location
        self.continue_on_error = continue_on_error

    def __call__(self, ics_file, org_file):
        try:
            cal = Calendar.from_ical(ics_file.read())
        except ValueError as e:
            msg = "Parsing error: {}".format(e)
            raise IcalError(msg)

        now = datetime.now(utc)
        start = now - timedelta(days=self.days)
        end = now + timedelta(days=self.days)
        for comp in recurring_ical_events.of(
            cal, keep_recurrence_attributes=True
        ).between(start, end):
            if event_is_declined(comp, self.emails):
                continue
            try:
                org_file.write(self.create_entry(comp))
            except Exception:
                print("Exception when processing:\n", file=sys.stderr)
                print(comp.to_ical().decode('utf-8') + "\n", file=sys.stderr)
                if self.continue_on_error:
                    print(traceback.format_exc(), file=sys.stderr)
                else:
                    raise

    def create_entry(self, comp):
        output = []
        summary = None
        if "SUMMARY" in comp:
            summary = comp['SUMMARY'].to_ical().decode("utf-8")
            summary = summary.replace('\\,', ',')
        location = None
        if "LOCATION" in comp:
            location = comp['LOCATION'].to_ical().decode("utf-8")
            location = location.replace('\\,', ',')
        if not any((summary, location)):
            summary = u"(No title)"
        else:
            summary += " - " + location if location and self.include_location else ''
        rec_event = "RRULE" in comp
        description = None
        if 'DESCRIPTION' in comp:
            description = '\n'.join(comp['DESCRIPTION'].to_ical()
                                    .decode("utf-8").split('\\n'))
            description = description.replace('\\,', ',')

        output.append(u"* {}".format(summary))
        if rec_event and self.RECUR_TAG:
            output.append(u" {}".format(self.RECUR_TAG))
        output.append(u"\n")

        # Get start/end/duration
        ev_start = None
        ev_end = None
        duration = None
        if "DTSTART" in comp:
            ev_start = comp["DTSTART"].dt
        if "DTEND" in comp:
            ev_end = comp["DTEND"].dt
            if ev_start is not None:
                duration = ev_end - ev_start
        elif "DURATION" in comp:
            duration = comp["DURATION"].dt
            if ev_start is not None:
                ev_end = ev_start + duration

        # Special case for some calendars that include times at midnight for
        # whole day events
        if isinstance(ev_start, datetime) and isinstance(ev_end, datetime):
            if ev_start.time() == MIDNIGHT and ev_end.time() == MIDNIGHT:
                ev_start = ev_start.date()
                ev_end = ev_end.date()

        # Format date/time appropriately
        if isinstance(ev_start, datetime):
            # Normal event with start and end
            output.append("  {}--{}\n".format(
                org_datetime(ev_start, self.tz), org_datetime(ev_end, self.tz)
                ))
        elif isinstance(ev_start, date):
            if ev_start == ev_end - timedelta(days=1):
                # single day event
                output.append("  {}\n".format(org_date(ev_start, self.tz)))
            else:
                # multiple day event
                output.append(
                    "  {}--{}\n".format(
                        org_date(ev_start, self.tz),
                        org_date(ev_end - timedelta(days=1), self.tz),
                    ))

        if description:
            output.append(u"{}\n".format(description))
        output.append(u"\n")
        return ''.join(output)


def check_timezone(ctx, param, value):
    if (value is None) or (value in all_timezones):
        return value
    click.echo(u"Invalid timezone value {value}.".format(value=value))
    click.echo(u"Use --print-timezones to show acceptable values.")
    ctx.exit(1)


def print_timezones(ctx, param, value):
    if not value or ctx.resilient_parsing:
        return
    for tz in all_timezones:
        click.echo(tz)
    ctx.exit()


@click.command(context_settings={"help_option_names": ['-h', '--help']})
@click.option(
    "--print-timezones",
    "-p",
    is_flag=True,
    callback=print_timezones,
    is_eager=True,
    expose_value=False,
    help="Print acceptable timezone names and exit.")
@click.option(
    "--email",
    "-e",
    multiple=True,
    default=None,
    help="User email address (used to deal with declined events). You can write multiple emails with as many -e options as you like.")
@click.option(
    "--days",
    "-d",
    default=90,
    type=click.IntRange(0, clamp=True),
    help=("Window length in days (left & right from current time. Default is 90 days). "
          "Has to be positive."))
@click.option(
    "--timezone",
    "-t",
    default=None,
    callback=check_timezone,
    help="Timezone to use. (Local timezone by default).")
@click.option(
    "--location/--no-location",
    "include_location",
    default=True,
    help="Include the location (if present) in the headline. (Location is included by default).")
@click.option(
    "--continue-on-error",
    default=False,
    is_flag=True,
    help="Pass this to attempt to continue even if some events are not handled",
)
@click.argument("ics_file", type=click.File("r", encoding="utf-8"))
@click.argument("org_file", type=click.File("w", encoding="utf-8"))
def main(ics_file, org_file, email, days, timezone, include_location, continue_on_error):
    """Convert ICAL format into org-mode.

    Files can be set as explicit file name, or `-` for stdin or stdout::

        $ ical2orgpy in.ical out.org

        $ ical2orgpy in.ical - > out.org

        $ cat in.ical | ical2orgpy - out.org

        $ cat in.ical | ical2orgpy - - > out.org
    """
    convertor = Convertor(days=days, tz=timezone,
                          emails=email, include_location=include_location,
                          continue_on_error=continue_on_error)
    try:
        convertor(ics_file, org_file)
    except IcalError as e:
        click.echo(str(e), err=True)
        raise click.Abort()


ical2orgpy/__init__.py
from __future__ import print_function

import sys
import traceback
from datetime import date, datetime, timedelta, time

import click
import recurring_ical_events
from icalendar import Calendar
from pytz import all_timezones, timezone, utc
from tzlocal import get_localzone

MIDNIGHT = time(0, 0, 0)

__version__ = "0.5"

def org_datetime(dt, tz):
    '''Timezone aware datetime to YYYY-MM-DD DayofWeek HH:MM str in localtime.
    '''
    return dt.astimezone(tz).strftime("<%Y-%m-%d %a %H:%M>")


def org_date(dt, tz):
    '''Timezone aware date to YYYY-MM-DD DayofWeek in localtime.
    '''
    if hasattr(dt, "astimezone"):
        dt = dt.astimezone(tz)
    return dt.strftime("<%Y-%m-%d %a>")


def event_is_declined(comp, emails):
    attendee_list = comp.get('ATTENDEE', None)
    if attendee_list:
        if not isinstance(attendee_list, list):
            attendee_list = [attendee_list]
        for att in attendee_list:
            if att.params.get('PARTSTAT', '') == 'DECLINED' and att.params.get('CN', '') in emails:
                return True
    return False


class IcalError(Exception):
    pass


class Convertor():
    RECUR_TAG = ":RECURRING:"

    # Do not change anything below

    def __init__(self, days=90, tz=None, emails=[], include_location=True, continue_on_error=False):
        """
        days: Window length in days (left & right from current time). Has
        to be positive.
        tz: timezone. If None, use local timezone.
        emails: list of user email addresses (to deal with declined events)
        """
        self.emails = set(emails)
        self.tz = timezone(tz) if tz else get_localzone()
        self.days = days
        self.include_location = include_location
        self.continue_on_error = continue_on_error

    def __call__(self, ics_file, org_file):
        try:
            cal = Calendar.from_ical(ics_file.read())
        except ValueError as e:
            msg = "Parsing error: {}".format(e)
            raise IcalError(msg)

        now = datetime.now(utc)
        start = now - timedelta(days=self.days)
        end = now + timedelta(days=self.days)
        for comp in recurring_ical_events.of(
            cal, keep_recurrence_attributes=True
        ).between(start, end):
            if event_is_declined(comp, self.emails):
                continue
            try:
                org_file.write(self.create_entry(comp))
            except Exception:
                print("Exception when processing:\n", file=sys.stderr)
                print(comp.to_ical().decode('utf-8') + "\n", file=sys.stderr)
                if self.continue_on_error:
                    print(traceback.format_exc(), file=sys.stderr)
                else:
                    raise

    def create_entry(self, comp):
        output = []
        summary = None
        if "SUMMARY" in comp:
            summary = comp['SUMMARY'].to_ical().decode("utf-8")
            summary = summary.replace('\\,', ',')
        location = None
        if "LOCATION" in comp:
            location = comp['LOCATION'].to_ical().decode("utf-8")
            location = location.replace('\\,', ',')
        if not any((summary, location)):
            summary = u"(No title)"
        else:
            summary += " - " + location if location and self.include_location else ''
        rec_event = "RRULE" in comp
        description = None
        if 'DESCRIPTION' in comp:
            description = '\n'.join(comp['DESCRIPTION'].to_ical()
                                    .decode("utf-8").split('\\n'))
            description = description.replace('\\,', ',')

        output.append(u"* {}".format(summary))
        if rec_event and self.RECUR_TAG:
            output.append(u" {}".format(self.RECUR_TAG))
        output.append(u"\n")

        # Get start/end/duration
        ev_start = None
        ev_end = None
        duration = None
        if "DTSTART" in comp:
            ev_start = comp["DTSTART"].dt
        if "DTEND" in comp:
            ev_end = comp["DTEND"].dt
            if ev_start is not None:
                duration = ev_end - ev_start
        elif "DURATION" in comp:
            duration = comp["DURATION"].dt
            if ev_start is not None:
                ev_end = ev_start + duration

        # Special case for some calendars that include times at midnight for
        # whole day events
        if isinstance(ev_start, datetime) and isinstance(ev_end, datetime):
            if ev_start.time() == MIDNIGHT and ev_end.time() == MIDNIGHT:
                ev_start = ev_start.date()
                ev_end = ev_end.date()

        # Format date/time appropriately
        if isinstance(ev_start, datetime):
            # Normal event with start and end
            output.append("  {}--{}\n".format(
                org_datetime(ev_start, self.tz), org_datetime(ev_end, self.tz)
                ))
        elif isinstance(ev_start, date):
            if ev_start == ev_end - timedelta(days=1):
                # single day event
                output.append("  {}\n".format(org_date(ev_start, self.tz)))
            else:
                # multiple day event
                output.append(
                    "  {}--{}\n".format(
                        org_date(ev_start, self.tz),
                        org_date(ev_end - timedelta(days=1), self.tz),
                    ))

        if description:
            output.append(u"{}\n".format(description))
        output.append(u"\n")
        return ''.join(output)


def check_timezone(ctx, param, value):
    if (value is None) or (value in all_timezones):
        return value
    click.echo(u"Invalid timezone value {value}.".format(value=value))
    click.echo(u"Use --print-timezones to show acceptable values.")
    ctx.exit(1)


def print_timezones(ctx, param, value):
    if not value or ctx.resilient_parsing:
        return
    for tz in all_timezones:
        click.echo(tz)
    ctx.exit()


@click.command(context_settings={"help_option_names": ['-h', '--help']})
@click.option(
    "--print-timezones",
    "-p",
    is_flag=True,
    callback=print_timezones,
    is_eager=True,
    expose_value=False,
    help="Print acceptable timezone names and exit.")
@click.option(
    "--email",
    "-e",
    multiple=True,
    default=None,
    help="User email address (used to deal with declined events). You can write multiple emails with as many -e options as you like.")
@click.option(
    "--days",
    "-d",
    default=90,
    type=click.IntRange(0, clamp=True),
    help=("Window length in days (left & right from current time. Default is 90 days). "
          "Has to be positive."))
@click.option(
    "--timezone",
    "-t",
    default=None,
    callback=check_timezone,
    help="Timezone to use. (Local timezone by default).")
@click.option(
    "--location/--no-location",
    "include_location",
    default=True,
    help="Include the location (if present) in the headline. (Location is included by default).")
@click.option(
    "--continue-on-error",
    default=False,
    is_flag=True,
    help="Pass this to attempt to continue even if some events are not handled",
)
@click.argument("ics_file", type=click.File("r", encoding="utf-8"))
@click.argument("org_file", type=click.File("w", encoding="utf-8"))
@click.version_option(__version__)
def main(ics_file, org_file, email, days, timezone, include_location, continue_on_error):
    """Convert ICAL format into org-mode.

    Files can be set as explicit file name, or `-` for stdin or stdout::

        $ ical2orgpy in.ical out.org

        $ ical2orgpy in.ical - > out.org

        $ cat in.ical | ical2orgpy - out.org

        $ cat in.ical | ical2orgpy - - > out.org
    """
    convertor = Convertor(days=days, tz=timezone,
                          emails=email, include_location=include_location,
                          continue_on_error=continue_on_error)
    try:
        convertor(ics_file, org_file)
    except IcalError as e:
        click.echo(str(e), err=True)
        raise click.Abort()
